### PR TITLE
Alibaba provider types update, consolidation, and clean up

### PR DESCRIPTION
**Pull request by kwoodson**
I have attempted to clean up the Alibaba provider types for preparation with the move to openshift/api. These changes have been recommended by the cloud provider team.

I will move forward with testing these changes.

cc @menglingwei @gujingit

### CODE REVIEW

File: pkg/actuators/machine/machine_scope.go [hunk 1]
```diff
@@ -24,13 +24,13 @@ import (
...
-	alibabacloudproviderv1 "github.com/AliyunContainerService/cluster-api-provider-alibabacloud/pkg/apis/alibabacloudprovider/v1beta1"
+	alibabacloudproviderv1 "github.com/openshift/cluster-api-provider-alibaba/pkg/apis/alibabacloudprovider/v1beta1"
 	corev1 "k8s.io/api/core/v1"
 	"k8s.io/klog"
 
-	v1beta1 "github.com/AliyunContainerService/cluster-api-provider-alibabacloud/pkg/apis/alibabacloudprovider/v1beta1"
-	alibabacloudClient "github.com/AliyunContainerService/cluster-api-provider-alibabacloud/pkg/client"
-	machinev1 "github.com/openshift/machine-api-operator/pkg/apis/machine/v1beta1"
+	machinev1 "github.com/openshift/api/machine/v1beta1"
+	v1beta1 "github.com/openshift/cluster-api-provider-alibaba/pkg/apis/alibabacloudprovider/v1beta1"
```
Comment (JoelSpeed) [inline pkg/actuators/machine/machine_scope.go:32]: Idiomatically we would call this `alibabav1`, but I guess as it's already v1beta1 we can leave as that for now, once it's migrated we will have to switch all the imports to `machinev1` anyway

File: pkg/apis/alibabacloudprovider/v1beta1/alibabacloudmachineproviderconfig_types.go [hunk 2]
```diff
@@ -58,186 +62,114 @@ type AlibabaCloudMachineProviderConfig struct {
...
+	// maximum number of security groups to which an instance can belong. For more information, see the "Security group limits" section in Limits.
+	// https://www.alibabacloud.com/help/doc-detail/101348.htm?spm=a2c63.p38356.879954.48.78f0199aX3dfIE
+	// +optional
+	SecurityGroups []ResourceTagReference `json:"securityGroups,omitempty"`
 
-	//The description of the system disk. The description must be 2 to 256 characters in length and cannot start with http:// or https://.
-	//
-	//This parameter is empty by default.
-	SystemDiskDescription string `json:"systemDiskDescription,omitempty"`
+	//Bandwidth describes the internet bandwidth strategy for the instance
```
Comment (JoelSpeed) [inline pkg/apis/alibabacloudprovider/v1beta1/alibabacloudmachineproviderconfig_types.go:71]: Nit, missing space
```suggestion
	// Bandwidth describes the internet bandwidth strategy for the instance
```

File: pkg/apis/alibabacloudprovider/v1beta1/alibabacloudmachineproviderconfig_types.go [hunk 3]
```diff
@@ -58,186 +62,114 @@ type AlibabaCloudMachineProviderConfig struct {
...
+	// +optional
+	Bandwidth BandwidthProperties `json:"bandwidth,omitempty"`
 
-	DataDisks []DataDisk `json:"dataDisk,omitempty"`
+	//Information regarding the the system disk for the instance
```
Comment (JoelSpeed) [inline pkg/apis/alibabacloudprovider/v1beta1/alibabacloudmachineproviderconfig_types.go:75]: Nit missing space, should follow golang convention of starting with the name
```suggestion
	// SystemDisk is the configuration for the instance system disk
```

File: pkg/apis/alibabacloudprovider/v1beta1/alibabacloudmachineproviderconfig_types.go [hunk 4]
```diff
@@ -58,186 +62,114 @@ type AlibabaCloudMachineProviderConfig struct {
...
+	// +optional
+	SystemDisk SystemDiskProperties `json:"systemDisk,omitempty"`
 
-	//The description of the instance. The description must be 2 to 256 characters in length and cannot start with http:// or https://.
-	//
-	//This parameter is empty by default.
-	Description string `json:"description,omitempty"`
+	//Information regarding the data disks for the instance
```
Comment (JoelSpeed) [inline pkg/apis/alibabacloudprovider/v1beta1/alibabacloudmachineproviderconfig_types.go:79]:
```suggestion
	// DataDisks is the configuration for additional disks attached to the instance
```

File: pkg/apis/alibabacloudprovider/v1beta1/alibabacloudmachineproviderconfig_types.go [hunk 5]
```diff
@@ -58,186 +62,114 @@ type AlibabaCloudMachineProviderConfig struct {
...
+	// +optional
+	DataDisks []DataDisk `json:"dataDisks,omitempty"`
 
 	//The ID of the vSwitch to which to connect the instance. This parameter is required when you create an instance of the VPC type. You can call the DescribeVSwitches operation to query the created vSwitches.
+	// +optional
 	VSwitchID string `json:"vSwitchId,omitempty"`
 
+	//VSwitch is a reference to the vswitch to use for this instance
+	//This parameter is required when you create an instance of the VPC type.
+	//You can call the DescribeVSwitches operation to query the created vSwitches.
```
Comment (JoelSpeed) [inline pkg/apis/alibabacloudprovider/v1beta1/alibabacloudmachineproviderconfig_types.go:89]:
```suggestion
	//. VSwitch is a reference to the vswitch to use for this instance
	// This parameter is required when you create an instance of the VPC type.
	// You can call the DescribeVSwitches operation to query the created vSwitches.
```

File: pkg/apis/alibabacloudprovider/v1beta1/alibabacloudmachineproviderconfig_types.go [hunk 6]
```diff
@@ -58,186 +62,114 @@ type AlibabaCloudMachineProviderConfig struct {
...
+	// +optional
 	Tenancy InstanceTenancy `json:"tenancy,omitempty"`
 
 	//The ID of the dedicated host on which to create the instance.
-	//
 	//You can call the DescribeDedicatedHosts operation to query the dedicated host list.
-	//
 	//When the DedicatedHostID parameter is specified, the SpotStrategy and SpotPriceLimit parameters are ignored. This is because preemptible instances cannot be created on dedicated hosts.
+	// +optional
 	DedicatedHostID string `json:"dedicatedHostId"`
```
Comment (JoelSpeed) [inline pkg/apis/alibabacloudprovider/v1beta1/alibabacloudmachineproviderconfig_types.go:147]: We should check how we do this on AWS, wondering if we group this into a struct there 🤔

File: pkg/apis/alibabacloudprovider/v1beta1/alibabacloudmachineproviderconfig_types.go [hunk 7]
```diff
@@ -247,6 +179,177 @@ type AlibabaCloudMachineProviderConfigList struct {
...
+	//Valid values when DataDisk.N.Category is set to cloud: 5 to 2000
+	//The value of this parameter must be greater than or equal to the size of the snapshot specified by the SnapshotID parameter.
+	// +optional
+	Size int `name:"size,omitempty"`
+
+	//Specifies whether to encrypt data disk N.
+	//
+	//Default value: false.
+	// +optional
+	Encrypted bool `name:"encrypted,omitempty"`
```
Comment (JoelSpeed) [inline pkg/apis/alibabacloudprovider/v1beta1/alibabacloudmachineproviderconfig_types.go:250]: Can the system disk be encrypted? We don't have a parameter for it right now

File: pkg/apis/alibabacloudprovider/v1beta1/alibabacloudmachineproviderconfig_types.go [hunk 8]
```diff
@@ -247,6 +179,177 @@ type AlibabaCloudMachineProviderConfigList struct {
...
+	Size int `json:"size,omitempty"`
+
+	//The description of the system disk. The description must be 2 to 256 characters in length and cannot start with http:// or https://.
+	//This parameter is empty by default.
+	// +optional
+	Description string `json:"description,omitempty"`
+}
+
+// DataDisk The datadisk of Instance
+type DataDisk struct {
```
Comment (JoelSpeed) [inline pkg/apis/alibabacloudprovider/v1beta1/alibabacloudmachineproviderconfig_types.go:222]: Wondering if a `SystemDisk` is actually a `DataDisk` as well, are all of these params applicable to both?

File: pkg/apis/alibabacloudprovider/v1beta1/alibabacloudmachineproviderconfig_types.go [hunk 9]
```diff
@@ -247,6 +179,177 @@ type AlibabaCloudMachineProviderConfigList struct {
...
+
+	//Specifies whether to release data disk N along with the instance.
+	//
+	//Default value: true.
+	// +optional
+	DeleteWithInstance *bool `name:"deleteWithInstance,omitempty"`
+}
+
+// Subscription information for the instance
+type SubscriptionInfo struct {
```
Comment (JoelSpeed) [inline pkg/apis/alibabacloudprovider/v1beta1/alibabacloudmachineproviderconfig_types.go:298]: Out of interest, not to do with the API, but what happens if the subscription for an instance runs out? Will Alibaba terminate the instance?

**Review by JoelSpeed — COMMENTED** [review]
I haven't highlighted all the nits (missing spaces, comment should start with var name) across the API file, but if we could go through and update it, this is important to do before we move to openshift/api

This is looking better though, I think this is a good improvement to the UX

File: pkg/apis/alibabacloudprovider/v1beta1/alibabacloudmachineproviderconfig_types.go (hunk 9 above)
Comment (kwoodson) [reply pkg/apis/alibabacloudprovider/v1beta1/alibabacloudmachineproviderconfig_types.go:298]: @menglingwei Could you provide some insight on this? What happens when subscriptions expire?

Comment (menglingwei) [reply pkg/apis/alibabacloudprovider/v1beta1/alibabacloudmachineproviderconfig_types.go:298]: @kwoodson when subscriptions are expired. The ecs instance will be stopped. If you do not renew , the instance will be released after some days(15days or more ,i need to double check it)

Comment (menglingwei) [reply pkg/apis/alibabacloudprovider/v1beta1/alibabacloudmachineproviderconfig_types.go:298]: @kwoodson  For more informations about subscriptions ,you can link to [help documentions](https://www.alibabacloud.com/help/doc-detail/56220.htm?spm=a2c63.p38356.879954.5.5ad77a56lYhtIg#subs-china)

Comment (menglingwei) [reply pkg/apis/alibabacloudprovider/v1beta1/alibabacloudmachineproviderconfig_types.go:298]: image

Comment (menglingwei) [reply pkg/apis/alibabacloudprovider/v1beta1/alibabacloudmachineproviderconfig_types.go:298]: @kwoodson  In OCP,which billing method will provide for end users ?

**Review by elmiko — COMMENTED** [review]
this generally looks good to me, are there any updates regarding @JoelSpeed 's questions/comments?

**Comment by kwoodson** [comment]
@elmiko I am updating this PR now. I have changed the tag types in #12 at Joel's request. I now have to take those changes and put them here. I'll get this cleaned up and ready. The question is whether it goes here first and then to openshift/api or remove it here and push the changes here to openshift/api?

**Comment by elmiko** [comment]
> The question is whether it goes here first and then to openshift/api or remove it here and push the changes here to openshift/api?

good question, my hunch is that we need to do the openshift/api merge, and then fix things up here. but i might be missing a detail.

File: pkg/apis/alibabacloudprovider/v1beta1/alibabacloudmachineproviderconfig_types.go (hunk 6 above)
Comment (kwoodson) [reply pkg/apis/alibabacloudprovider/v1beta1/alibabacloudmachineproviderconfig_types.go]: I didn't see a dedicated host ID. Only information regarding tenancy.

File: pkg/apis/alibabacloudprovider/v1beta1/alibabacloudmachineproviderconfig_types.go (hunk 7 above)
Comment (kwoodson) [reply pkg/apis/alibabacloudprovider/v1beta1/alibabacloudmachineproviderconfig_types.go:247]: @menglingwei Is it possible to encrypt a system disk?

File: pkg/apis/alibabacloudprovider/v1beta1/alibabacloudmachineproviderconfig_types.go (hunk 8 above)
Comment (kwoodson) [reply pkg/apis/alibabacloudprovider/v1beta1/alibabacloudmachineproviderconfig_types.go]: It appears that the `SystemDisk` and the `DataDisk` types have different properties.
https://github.com/aliyun/alibaba-cloud-sdk-go/blob/1c273f0a7d673a8f941f67cfe572c0a3ea1a7122/services/ecs/struct_data_disk.go
https://github.com/aliyun/alibaba-cloud-sdk-go/blob/64043b70c5e4a928281b4114300481e027111abb/services/ens/struct_system_disk.go

File: pkg/apis/alibabacloudprovider/v1beta1/alibabacloudmachineproviderconfig_types.go (hunk 9 above)
Comment (kwoodson) [reply pkg/apis/alibabacloudprovider/v1beta1/alibabacloudmachineproviderconfig_types.go:295]: @menglingwei I'm not sure which method. What do you recommend?

File: pkg/apis/alibabacloudprovider/v1beta1/alibabacloudmachineproviderconfig_types.go (hunk 2 above)
Comment (kwoodson) [reply pkg/apis/alibabacloudprovider/v1beta1/alibabacloudmachineproviderconfig_types.go:70]: Fixed.

File: pkg/apis/alibabacloudprovider/v1beta1/alibabacloudmachineproviderconfig_types.go (hunk 3 above)
Comment (kwoodson) [reply pkg/apis/alibabacloudprovider/v1beta1/alibabacloudmachineproviderconfig_types.go]: Fixed

File: pkg/apis/alibabacloudprovider/v1beta1/alibabacloudmachineproviderconfig_types.go (hunk 4 above)
Comment (kwoodson) [reply pkg/apis/alibabacloudprovider/v1beta1/alibabacloudmachineproviderconfig_types.go]: Fixed

File: pkg/apis/alibabacloudprovider/v1beta1/alibabacloudmachineproviderconfig_types.go (hunk 5 above)
Comment (kwoodson) [reply pkg/apis/alibabacloudprovider/v1beta1/alibabacloudmachineproviderconfig_types.go]: fixed

**Comment by kwoodson** [comment]
Addressed the feedback from Joel on spacing. I'll plan on a merge here. Then a merge to openshift/api and a subsequent removal here.

File: pkg/apis/alibabacloudprovider/v1beta1/alibabacloudmachineproviderconfig_types.go (hunk 9 above)
Comment (menglingwei) [reply pkg/apis/alibabacloudprovider/v1beta1/alibabacloudmachineproviderconfig_types.go:297]: @kwoodson  I suggest pay-as-you-go.

**Review by elmiko — COMMENTED** [review]
/lgtm

**Pull request merged**